### ISSUE TITLE
Adopt new mechanism for publishing extension insiders builds

### ISSUE THREAD

**Issue by joyceerhl** [microsoft/vscode-remote-repositories-github#91]
See https://github.com/microsoft/vscode/issues/15756

**Comment by isidorn** [comment]
@joyceerhl if you need help feel free to reach out. I will write some docs later this week to help with this.

**Comment by isidorn** [comment]
Here are the docs  https://github.com/microsoft/vscode-docs/blob/vnext/api/working-with-extensions/publishing-extension.md#pre-release-extensions 
Soon they will be on the website.

**Comment by joyceerhl** [comment]
Problem
- Currently RemoteHub has Insiders and Stable, published under separate extension IDs
- RemoteHub wants to adopt the new mechanism for publishing pre-releases https://github.com/microsoft/vscode/issues/15756
- RemoteHub stores uncommitted changes locally in extension memento storage
- We need to ensure that uncommitted changes don't get lost when RemoteHub transitions to publishing pre-release extensions
- **The worst case scenario is data loss for RemoteHub Insiders users**

Who is using RemoteHub Insiders
- Currently insiders.vscode.dev bundles RemoteHub Insiders, stable vscode.dev bundles RemoteHub Stable
- Desktop users may also install RemoteHub Insiders
- github.dev always bundles RemoteHub Stable. Are there any users who have somehow managed to install RemoteHub Insiders in github.dev?

Case study: GHPRI
- Extension auto updated but extension state not migrated
- Extension doesn't get published anymore
- Extension marketplace page says that it's now part of the pre release publishing process

Other migrations we have done in the past
- Multiroot workspace migration
- Migration from local storage to a SQLite database https://github.com/microsoft/vscode/issues/61928

Corner cases to consider
- Issues with writes from multiple windows? https://github.com/microsoft/vscode/issues/125970 Browser analogue: what if the user has the same workspace open in multiple tabs, or multiple insiders.vscode.dev windows open simultaneously?
- What if the migration is partial / interrupted? E.g. user closes the tab 
    - SQLite is good at recovering, how about IndexedDB?  https://github.com/microsoft/vscode/issues/116841
    - And how do we tell that the migration is complete? E.g. memento storage used the lowercase heuristic but that turned out to not be accurate --> rename the keys foo to foo-migrated or similar when migration is complete?
    - What is the average size of uncommitted changes?
- Incompatible data formats https://github.com/microsoft/vscode/issues/102001 (should not be an issue for now but could become a problem eventually e.g. if RemoteHub changes the format of the data that we store, we need to ensure that we continue being able to read stuff out)

Proposed user migration experience
- insiders.vscode.dev bundles the pre-release version of RemoteHub
    - The first time users access the insiders.vscode.dev/github route, they will get a modal dialog informing them that data needs to be migrated 
        - "We need to perform a one-time migration of your local uncommitted changes. Warning: if you do not migrate your uncommitted changes now, you will not be able to access them in future."
        - With options "Begin Migration", "I Understand, Do Not Migrate"
        - Clicking 'x' should be treated as a 'Not Now' and we will continue to prompt to migrate in future
        - While migration is taking place, we should display a progress notification (maybe not necessary if there are very few changes to migrate?) "Migration in progress. Please do not reload the window."
        -  Problem is that most users have no idea what RemoteHub is, especially since it does not show up as an installed extension in the extension marketplace panel given that it is bundled as builtin by the workbench embedder, so this language is probably going to be confusing
        - Alternative: maybe we just immediately start the migration? I would rather not, because I don't know what's going to happen if the user bounces from the page while a migration is ongoing (and the user is much more likely to do this in web than in desktop)
- On desktop, RemoteHub is *not* bundled. When RemoteHub insiders activates, the user will also get a modal dialog informing them that data needs to be migrated
        - The dialog *should* explicitly mention the RemoteHub Insiders extension being deprecated in favor of the pre-release, because the user explicitly opted into the extension
        - Instead of 'Begin Migration', it will include an option to "Switch to Pre-release and Migrate My Uncommitted Changes", which will install RemoteHub pre-release and copy uncommitted changes to where RemoteHub expects to read them (NB: There is no need to include the option to switch to pre-release in insiders.vscode.dev, because it's bundled by default, but there we should still show a modal so that the user understands that they should not close or reload the tab while the migration is taking place)
- Do not delete old data, but store a key in memento storage for the stable extension to indicate that migration from insiders took place on this client

Open questions
- Make sure github.dev is aware / not impacted by proposed migration
- Insiders changes will now be visible from stable since we are just copying the memento storage sideways, is that a problem or should we implement some scoping? (Personally I think it's desirable that changes which lived in insiders are now available from the stable quality, and that is part of my proposal for roaming uncommitted changes)
- Do we need to migrate other extension data besides uncommitted changes? What else does RemoteHub store in local storage?
- How to resolve merge conflicts e.g. if the same workspace has uncommitted changes in both stable and insiders?

**Comment by joyceerhl** [comment]
Another thought: today we bundle the insiders version of RemoteHub in insiders.vscode.dev through the `builtinExtensions` workbench construction option but I don't imagine we can do that via the workbench construction options once RemoteHub adopts the prerelease publishing flow. We could just let the user manually switch to the pre-release version even in insiders.vscode.dev. We would need to ensure that the prerelease extension flow is smooth for extensions which are bundled as builtin in the workbench embedder, cc @joaomoreno

**Comment by joaomoreno** [comment]
> Another thought: today we bundle the insiders version of RemoteHub in insiders.vscode.dev through the `builtinExtensions` workbench construction option but I don't imagine we can do that via the workbench construction options once RemoteHub adopts the prerelease publishing flow. We could just let the user manually switch to the pre-release version even in insiders.vscode.dev. We would need to ensure that the prerelease extension flow is smooth for extensions which are bundled as builtin in the workbench embedder, cc @joaomoreno

What do you think we should do, @sandy081?

**Comment by joaomoreno** [comment]
Oh I guess we should just follow the same as https://github.com/microsoft/vscode-dev/issues/467.

**Comment by sandy081** [comment]
Yes. Same as GHPRI extension.

- Currently Workbench automatically migrates to GHPRI pre-release version when embedder provides GHPRI nightly extension
- You can start passing GHPRI pre-release version using the new following embedder api 
 
```
/**
 * The identifier of an extension in the format: `PUBLISHER.NAME`.
 * For example: `vscode.csharp`
 */
type ExtensionId = string;

type MarketplaceExtension = ExtensionId | { readonly id: ExtensionId, preRelease?: boolean };
/**
 * Additional builtin extensions those cannot be uninstalled but only be disabled.
 * It can be one of the following:
 * 	- an extension in the Marketplace
 * 	- location of the extension where it is hosted.
 */
readonly additionalBuiltinExtensions?: readonly (MarketplaceExtension | UriComponents)[];
```

Please let me know if there is any help I can do or if you need any info.

**Comment by sandy081** [comment]
Automatic migration does only uninstalling the nightly extension and install the pre-release extension. Storage is not migrated. It is mentioned in the Note here https://github.com/microsoft/vscode-docs/blob/vnext/api/working-with-extensions/publishing-extension.md#pre-release-extensions

**Comment by isidorn** [comment]
We have internal commands to install pre-release extension and to uninstall extensions, so if you would like to customise this experience you could do it from the Remote-Repositories extension. For example show a dialog, notification and only once user agrees you execute those commands...

**Comment by sandy081** [comment]
This might not work for GitHub Repositories extension in insiders.vscode.dev because it is an additional builtin extension contributed by vscode.dev. So even though extension uninstalls the nightly extension, the nightly extension will come back on reload.

**Comment by joyceerhl** [comment]
After talking to Sandeep yesterday about a related problem, I now believe that RemoteHub can adopt pre-releases the same way that GHPRI did, i.e.:
- Copy github.remotehub-insiders extension storage to github-remotehub extension storage on vscode.dev
- Avoid migrating github.remotehub-insiders on desktop, and instead simply mark it as deprecated 

## vscode.dev
On vscode.dev, RemoteHub and VS Code qualities are tightly coupled--RemoteHub Insiders is pre-installed in insiders.vscode.dev, and RemoteHub Stable is pre-installed in vscode.dev. There's no way to uninstall the bundled RemoteHub extension, and users will get a warning if they try to install RemoteHub Insiders in vscode.dev or RemoteHub Stable in insiders.vscode.dev. I learned from Sandeep that insiders.vscode.dev and vscode.dev have separate IndexDB instances, so we won't be overwriting any changes by copying github.remotehub-insiders extension storage to github.remotehub extension storage on vscode.dev.

![image](https://user-images.githubusercontent.com/30305945/145622755-e1a1096c-a7b6-40d4-bfdd-a25c2c2c41a1.png)

## github.dev
On github.dev, RemoteHub Stable is always pre-installed, regardless of VS Code quality. Any usage of RemoteHub Insiders results in an error notification. There shouldn't be any cases where RemoteHub Insiders is used. (Note however that it looks like github.dev has the same shared IndexDB instance across VS Code qualities)

## Desktop
On desktop, a user may have installed:
- RemoteHub Insiders and RemoteHub Stable in VS Code Stable
- RemoteHub Insiders and RemoteHub Stable in VS Code Insiders
In this situation, VS Code core cannot simply overwrite github.remotehub extension storage with github.remotehub-insiders extension storage. At minimum we would need to check which one the user would prefer before blowing away user data. In this scenario I think it might be better to simply mark the extension as deprecated but not uninstall it, and on extension activation, display a warning that says "This extension has been deprecated in favor of pre-release versions of the github.remotehub extension. Please commit all your uncommitted changes to avoid losing work."

![image](https://user-images.githubusercontent.com/30305945/145623315-cf215058-0739-4053-b22f-50573da8ee28.png)

@sandy081 let me know if I have this right. If so, then I believe we can proceed with the following action items:

Changes to VS Code core
- Do same thing that we did for GHPRI, i.e.
    - [ ] Migrate extension storage in web
    - [ ] Mark extension as deprecated

Changes to RemoteHub
- [x] Require VS Code >1.63 (we already do this due to adopting `enabledApiProposals`)
- [ ] Change `Publish insiders` workflow to publish with `--pre-release` flag instead of tacking on `-insider` to extension id
- [ ] After bumping version number to publish stable builds, also bump version number in main branch so that VS Code picks up pre-releases over stable releases if the user has opted into pre-releases
- [ ] Remove the error message that we display when github.remotehub-insiders and github.remotehub are installed together, since that will be redundant

**Comment by sandy081** [comment]
@joyceerhl Above makes sense to me. 

Please correct me if I understand correctly:

- You would want VS Code to migrate (install & storage) the extension in Web
- You will take care of uninstalling and installing pre-release version in Desktop

**Comment by joyceerhl** [comment]
On desktop we will have a notification indicating that `github-remotehub-insiders` is deprecated and asking the user to ensure they have committed any changes that they need, with two action buttons to 
1. Install Pre-Release
2. Don't Show Again

The Install Pre-Release option will install RemoteHub pre-releases and then uninstall RemoteHub Insiders.

Lad convinced me today that we should not use a modal dialog on desktop because it's up to the user if they choose to continue using the deprecated RemoteHub Insiders extension. We will also explicitly call out in the RemoteHub Insiders README that RemoteHub Insiders is deprecated and unsupported.

**Comment by joyceerhl** [comment]
Heres' the Insiders deprecation flow on desktop:
1. Deprecation warning notification and README are displayed on extension activation
![image](https://user-images.githubusercontent.com/30305945/146426289-001d94da-24da-45fe-98ca-d575d200a1dd.png)
2. Select the action button to install
![image](https://user-images.githubusercontent.com/30305945/146426295-c9c8e831-6d8d-4b27-a239-316b73725d59.png)
3. Insiders extension gets uninstalled, stable extension gets installed
![image](https://user-images.githubusercontent.com/30305945/146426300-903f0898-7960-4f78-92b9-e74bcf71148d.png)

The pre-release version of the extension will also include a notice saying that it works best with VS Code Insiders, which is what we had for RemoteHub Insiders builds:

![image](https://user-images.githubusercontent.com/30305945/146430155-178874d6-5568-4bb6-9a42-03522578a9d8.png)

**Comment by joyceerhl** [comment]
While testing migrating uncommitted changes from `github.remotehub-insiders` to the pre-release of `github.remotehub`, I discovered that RemoteHub actually stores uncommitted changes in two separate IndexedDB databases:
1. It writes metadata representing which files have uncommitted changes to the `vscode-web-state-db-global` database. This metadata is stored under the extension ID, i.e. `github.remotehub-insiders`. 
1. The actual changed file contents are stored in the `vscode-web-db` database via `vscode-userdata` URIs (this is abstracted away from RemoteHub, we store these using the `ExtensionContext.globalStorage` URI and `workspace.fs.writeFile`). If we don't migrate this data as well, the SCM view will say that files have changed, but RemoteHub won't be able to resolve the actual changed contents, so for example opening the diff editor will show that there are no uncommitted changes.
> ![image](https://user-images.githubusercontent.com/30305945/146596051-27ccb729-f6a5-4383-ab35-5a2f8612e0d5.png)

So to completely migrate uncommitted changes and ensure that they continue to be available in web, we would need to:
1. Copy the `github.remotehub-insiders` value in the `vscode-web-state-db-global` table to a new key, `github.remotehub`. I simulated this by running the following snippet in devtools:

> ```js
> var connection = indexedDB.open('vscode-web-state-db-global', 1);
> connection.onsuccess = (e) => {
>     var database = e.target.result;
>     var transaction = database.transaction('ItemTable', 'readwrite');
> 
>     var objectStore = transaction.objectStore('ItemTable');
> 
>     var cursor = objectStore.openCursor();
> 
>     const data = new Promise((resolve, reject) => {
>         cursor.onsuccess = () => {
>             if (cursor === undefined || cursor.result.key === 'github.remotehub-insiders') {
>                 resolve(cursor.result);
>             } else {
>                 cursor.result.continue();
>             }
>         }        
>     });
> 
>     data.then((d) => {
>         objectStore.put(d.value, 'github.remotehub');
>     });
> }
> ```

2. Copy any values stored under a key that begins with `/User/globalStorage/github.remotehub-insiders/` in the `vscode-web-db` table to a key that begins with `/User/globalStorage/github.remotehub/`. I simulated this in RemoteHub by doing a string replace on the `globalStorage` uri path that RemoteHub prefixes the URIs for all files with uncommitted changes with.

Extensions can't access other extensions' memento storage, so we definitely need workbench help with part 1. RemoteHub cannot enumerate the keys in the `vscode-web-db` table since that is abstracted away behind the `vscode-userdata` URI, so RemoteHub can't cleanly attempt part 2--we would have to do the migration as each file read request comes in. @sandy081, Is this something that the workbench can help to migrate as well?

**Comment by eamodio** [comment]
@joyceerhl in your comment above (https://github.com/microsoft/vscode-remote-repositories-github/issues/91#issuecomment-991201718) you mention that the Insiders version of RemoteHub can't be used on github.dev -- but it certainly can as I have been using it that way for a while (might require switching to the Insiders version of VS Code though).

And as for the desktop migration (or lack of) plan -- will it be clear to users what "data" would be lost? Because it could certainly be spread across many repositories (not just the currently opened one).

**Comment by joyceerhl** [comment]
@eamodio How did you get RemoteHub Insiders installed in github.dev with Insiders VS Code? With VS Code Insiders in github.dev, if I uninstall the stable build of RemoteHub and install the insiders build of RemoteHub, the stable build of RemoteHub comes back on a reload and I get the error notification from RemoteHub about the side by side stable/insiders installs not being supported. The only way this seems possible is if you disable the stable build of RemoteHub and install the insiders build of RemoteHub, and in that case it seems unlikely the user would have stable changes to retrieve. (And there's no option to disable the stable build of RemoteHub in vscode.dev--not sure why)

> will it be clear to users what "data" would be lost? Because it could certainly be spread across many repositories (not just the currently opened one)

We should definitely clearly call that out in the message we display to users, and in the extension README. Was there any language from the previous extension ID rename that we can reuse? We should also call out how users can retrieve their changes the way we did for `ms-vscode.remotehub`:

> ## Migrating from the previous (deprecated) version
> 
> If you were using the previous version of Remote Repositories, and have uncommitted changes that you want to retrieve, please install the [Remote Repositories (Deprecated) extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.remotehub). Then open the desired repository and commit your changes.

**Comment by eamodio** [comment]
@joyceerhl Yeah, you need to disable the Stable version and install the Insiders version. I agree, that if you are in that mode you likely don't have stable data that you care about, but we'd need to migrate the Insiders version of data -- so wanted to make sure that would happen.

**Comment by sandy081** [comment]
> Is this something that the workbench can help to migrate as well?

Currently I am working on migrating all extension storages from insiders extension to pre-release extension, this includes

- globalState (Mementos)
- workspaceState (Mementos)
- globalStorageUri (files stored in user data dir)
- storageUri (files stored in user data dir per workspace)

**Comment by sandy081** [comment]
@joyceerhl Summarizing my understandings. Please correct me if wrong

- In Desktop, you will guide users to uninstall the insiders extension and install the pre-release version. So workbench does not need to do anything here.
 
- In Web, what you are doing is not possible because the extension is a built in extension and workbench does not support uninstalling and installing built in extensions. Moreover you also need support for storage maigration, once you start passing main extension with pre-release flag through the `additionalBuiltinExtensions` embedder api in insiders.vscode.dev. 

https://github.com/microsoft/vscode/blob/5e630c145f53e9a9a2975806bf8530e64ba43ae5/src/vs/workbench/workbench.web.api.ts#L35

So, I would recommend following:

- Workbench will not do anything for your extension in Desktop
- In web, you can signal workbench (through a new embedder API) to migrate storage from one extension (your insiders extension) to new extension (your pre-release extension). Migration of storage includes following:
  - globalState (Mementos)
  - workspaceState (Mementos)
  - globalStorageUri (files stored in user data dir)
  - storageUri (files stored in user data dir per workspace)

Let me know if it covers your major scenarios?

**Comment by joyceerhl** [comment]
Yes, to my knowledge that should cover RemoteHub's needs for migration. To signal the workbench through API, would we have something like the following?

```ts
type MarketplaceExtension = ExtensionId | { readonly id: ExtensionId, preRelease?: boolean, migrateStorage?: boolean };
```

**Comment by sandy081** [comment]
Exactly ( you read my mind :) ), but with a small change 

```
type MarketplaceExtension = ExtensionId | { readonly id: ExtensionId, preRelease?: boolean, migrateStorageFrom?: ExtensionId };
```

**Comment by joyceerhl** [comment]
Yes that makes sense to me 😊 I have a PR to adopt the `preRelease` option for RemoteHub and GHPRI in the vscode.dev embedder, so when the `migrateStorageFrom` option is available in Insiders, I will update that PR to signal storage migration for RemoteHub and test to ensure that the migration works as expected.

**Comment by sandy081** [comment]
@joyceerhl Added above API - https://github.com/microsoft/vscode/issues/140088 - this will be available in tomorrow's insiders. Please test it out and adopt. Thanks.

**Comment by joyceerhl** [comment]
Tested API in Code-OSS + web, migration works like a charm 🙏Will test again with VS Code Insiders + web tomorrow before merging my vscode.dev PR. Then I'll publish a final Insiders update with the deprecation notice so that desktop users will be notified, and remove our RemoteHub Insiders publishing workflow.

**Comment by joyceerhl** [comment]
API looks good in Insiders too with some more testing this morning--now merging and deploying embedder adoption of `preRelease` and `migrateStorageFrom` flags for `additionalBuiltinExtensions`.

**Comment by joyceerhl** [comment]
There seems to be an issue where the global storage contents get wiped. I've reverted my change and redeployed vscode.dev for now so that more insiders.vscode.dev users do not run into this. From debugging with Sandeep we saw this in the window logs:

```
[2022-01-12 09:59:26.699] [window] [info] Migrating workspace extension storage from GitHub.remotehub-insiders to GitHub.remotehub...
[2022-01-12 09:59:26.837] [window] [info] Migrated workspace extension storage from GitHub.remotehub-insiders to GitHub.remotehub
[2022-01-12 09:59:26.957] [window] [info] Migrating global extension storage from github.remotehub-insiders to GitHub.remotehub...
[2022-01-12 09:59:27.021] [window] [info] Migrated global extension storage from github.remotehub-insiders to GitHub.remotehub
[2022-01-12 09:59:27.086] [window] [info] Migrating workspace extension storage from github.remotehub-insiders to GitHub.remotehub...
[2022-01-12 09:59:27.150] [window] [info] Migrated workspace extension storage from github.remotehub-insiders to GitHub.remotehub
```

**Comment by rchiodo** [comment]
/verified

Tried the deprecated version. It told me to update:

![image](https://user-images.githubusercontent.com/19672699/151259142-c10338e5-b471-4b11-a7e1-998f59a911a0.png)

Prerelease is default in insiders:

![image](https://user-images.githubusercontent.com/19672699/151259219-ac52cd34-0f15-453f-90d7-b443be41bfa0.png)
